### PR TITLE
fix vanishing toots bug when refreshing

### LINKED ISSUE #1228: Refresh issue when existing refresh running
When an existing background refresh is running (eg: just opened app, timed background refresh), and the user pulls down to refresh, you get issues with toots disappearing from the Home timeline. Doesn't always happen though, but the below should allow someone to reproduce somewhat reliably.

1. Start Tusky and wait for an update to finish, so it should have all the latest toots.
2. Quit Tusky (force kill if needed so it's not running in the background).
3. Log into the web version of the Mastodon instance and wait till 2-3 new toots have appeared in the Home timeline.
4. Start Tusky and within 2-3 seconds of the UI loading, pull down to refresh.
5. Compare this list of toots with the Mastodon web version. If they are different, forcing a change of the theme seems to restore the missing toots (eg: change from Black to Dark to Black again).

Notes:
- I suspect a workaround for this would be to not allow multiple refreshes to fire if a refresh is already happening (eg: some sort of lock around the refresh function). A status icon that shows when a refresh is happening would probably help as well (at least with further diagnosis).
- The issue can happen if you manage to pull down refresh as a background timed refresh has just started as well (not just at startup).

* * * *
- Tusky Version: 6.1
- Android Version: 8.0
- Android Device: Samsung SM-G930F
- Mastodon instance (if applicable): social.chinwag.org

- [*] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by connyduck**
closes #1228

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java [hunk 1]
```diff
@@ -305,22 +306,32 @@ private void updateCurrent() {
                             if (!statuses.isEmpty()) {
                                 filterStatuses(statuses);
 
-                                // Working around a bug when Mastodon API doesn't return the first
-                                // status because of string "id < maxId". Hacking with ID doesn't
-                                // help.
-                                if (!this.statuses.isEmpty()) {
-                                    Either<Placeholder, Status> firstOld = this.statuses.get(0);
-                                    this.statuses.clear();
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: So this line basically was the problem: it removes all statuses. But there might have been a concurrent top load going on in the meantime that inserted newer stauses. I fixed it by clearing only statuses that are older than the `topId` at the moment the load was started.

File: app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java [hunk 2]
```diff
@@ -284,6 +284,7 @@ private void tryCache() {
                     }
 
                     this.updateCurrent();
+                    this.loadAbove();
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: Now that concurrent refreshes and top loads are possible, lets to them in parallel.

File: app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java [hunk 3]
```diff
@@ -1171,7 +1183,7 @@ private void updateStatuses(List<Either<Placeholder, Status>> newStatuses, boole
 
     private void removeConsecutivePlaceholders() {
         for (int i = 0; i < statuses.size() - 1; i++) {
-            if (!statuses.get(i).isRight() && !statuses.get(i + 1).isRight()) {
+            if (statuses.get(i).isLeft() && statuses.get(i + 1).isLeft()) {
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: using `isLeft` is better than `!isRight`

File: app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java (hunk 1 above)
Comment (Tak) [reply app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: Is the mastodon bug mentioned in the comment still addressed now?

Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: I am pretty sure it is, but I will check again

Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: We should consider using new API parameters finally but not in this PR ofc

File: app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java [hunk 4]
```diff
@@ -305,22 +306,32 @@ private void updateCurrent() {
...
-                                    this.statuses.add(firstOld);
-                                } else {
-                                    this.statuses.clear();
+
+                                if (!this.statuses.isEmpty() && topId != null) {
+                                    // clear old cached statuses
+                                    Iterator<Either<Placeholder, Status>> iterator = statuses.iterator();
+                                    while (iterator.hasNext()) {
+                                        Either<Placeholder, Status> item = iterator.next();
+                                        if(item.isRight()) {
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: You could shorten it if you put id into the variable and left only one comparison

Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: Could also find the first newer one and delete everything older but not sure if there's any performance/readability benefit

File: app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java (hunk 3 above)
Comment (mal0ki) [reply app/src/main/java/com/keylesspalace/tusky/fragment/TimelineFragment.java]: :rofl:

**Comment by connyduck** [comment]
Tested it again, everything seems to work correctly

**Pull request merged**